Fix permissions for endpointslice controller

marun opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

When the `OwnerReferencesPermissionEnforcement` validating admission plugin is enabled, the EndpointSlice controller needs permission to update a service's finalizers to be able to create an EndpointSlice resource that is owned by the service and sets blockOwnerDeletion=true in its ownerRef.

The lack of this permission breaks the EndpointSlice controller on openshift, which enables `OwnerReferencesPermissionEnforcement` by default. The problem isn't apparent on kube due to `OwnerReferencesPermissionEnforcement` being disabled by default.

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed the EndpointSlice controller to run without error on a cluster with the OwnerReferencesPermissionEnforcement validating admission plugin enabled.
```

marun committed:
Fix permissions for endpointslice controller

The controller needs to be able to set a service's finalizers to be
able to create an EndpointSlice resource that is owned by the service
and sets blockOwnerDeletion=true in its ownerRef.

marun commented:
/assign @robscott @thockin

thockin reviewed (commented):
I don't know much about how blockOwnerDeletion is implemented - is that synchronous to the original REST call?

This seems OK, so I will approve, but Rob needs to LGTM

/approve

robscott commented:
@thockin I've been talking with @marun on Slack about this PR, and agree that it's a good fix. This was not noticed in k/k because we do not enable the `OwnerReferencesPermissionEnforcement` admission controller by default. When that is enabled it prevents the EndpointSlice controller from working with errors like this: `Error creating EndpointSlice for Service foo: endpointslices.discovery.k8s.io "foo-ttsrn" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on`. So hopefully not something that a lot of people will run into, but still very much worth fixing. Likely worth cherry-picking this into 1.18 and 1.17 as well. 

/lgtm

marun commented:
I've updated the PR description to reflect the involvement of the `OwnerReferencesPermissionEnforcement` admission plugin and added a release note that reflects this.

robscott commented:
/sig network
/priority important-soon

robscott commented:
/retest
